deal with intents on SystemDB

tbg opened the pull request:
The previous code did not handle these. Since a leftover intent on
this table may potentially only ever be touched by the code changed
in this PR, the previous version would stop gossipping the schema altogether.

With this change, when intents are encountered, an error is returned
but the intents are handled. This causes them to be cleaned up when
they are abandoned, but does not interfere with active writers. The
hope is that on the next iteration, the read can succeed. It is
going to be in our interest to minimize contention on that table.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3036)

<!-- Reviewable:end -->

tamird commented on storage/replica.go:
```diff
@@ -1573,14 +1573,24 @@ func (r *Replica) resolveIntents(ctx context.Context, intents []roachpb.Intent)
-	// TODO(tschottdorf): Currently this does not handle intents well.
-	kvs, _, err := engine.MVCCScan(eng, keys.SystemDBSpan.Key, keys.SystemDBSpan.EndKey,
-		0, roachpb.MaxTimestamp, true /* consistent */, nil)
+func loadSystemDBSpan(r *Replica) ([]roachpb.KeyValue, []byte, error) {
```
nit: make this a method?

tschottdorf replied:
done.

tamird commented on storage/replica.go:
```diff
@@ -1573,14 +1573,24 @@ func (r *Replica) resolveIntents(ctx context.Context, intents []roachpb.Intent)
+		// to nudge the intents in case they're expired; next time around we'll
+		// hopefully have more luck.
+		r.handleSkippedIntents(intents)
+		return nil, nil, util.Errorf("must retry later due to intent on SystemDB")
```
when is later? shouldn't we be able to retry immediately?

tschottdorf replied:
intent resolution is async, and the resolution here will only resolve "abandoned" intents. if a txn is still running, the intent will not be moved.

tamird commented on storage/replica_test.go:
```diff
@@ -3302,3 +3302,48 @@ func TestReplicaQuiesce(t *testing.T) {
+	if rng == nil {
+		t.Fatalf("no replica contains the SystemDB span")
+	}
+	v := roachpb.Value{}
```
nit: `roachpb.ValueFromBytes`

tschottdorf replied:
done.

tamird commented on storage/replica_test.go:
```diff
@@ -3302,3 +3302,48 @@ func TestReplicaQuiesce(t *testing.T) {
+	v.SetBytes([]byte("foo"))
+	// Create a transaction. We don't write a txn record, so pushing this will
+	// succeed and abort the intent we write here.
+	txn := newTransaction("test", []byte("a"), 1, roachpb.SNAPSHOT, rng.store.Clock())
```
Does it matter that this is SNAPSHOT?

tschottdorf replied:
nope, just had to pass something. Changing to SERIALIZABLE for least surprise.

tamird commented on storage/replica_test.go:
```diff
@@ -3302,3 +3302,48 @@ func TestReplicaQuiesce(t *testing.T) {
+		t.Fatal(err)
+	}
+	// Verify that this trips up loading the SystemDB data.
+	if _, _, err := loadSystemDBSpan(rng); !testutils.IsError(err, "must retry later due to intent") {
```
nit: extract this error into a "constant" you can match exactly

tschottdorf replied:
done.

tamird commented on storage/replica_test.go:
```diff
@@ -3302,3 +3302,48 @@ func TestReplicaQuiesce(t *testing.T) {
+	// In the loop, wait until the intent is aborted. Then write a "real" value
+	// there and verify that we can now load the data as expected.
+	util.SucceedsWithin(t, time.Second, func() error {
+		kvs, _, err := loadSystemDBSpan(rng)
```
the `loadSystemDBSpan` outside the loop already triggers the async cleanup, right? Is there any reason this needs to happen before the `Put` below?

tschottdorf replied:
no, you could move the Put up.

tamird replied:
let's do that; on first reading I was looking around for where `kvs` was being used in the Put below.

tschottdorf replied:
done.

tamird commented on storage/replica_test.go:
```diff
@@ -3302,3 +3302,48 @@ func TestReplicaQuiesce(t *testing.T) {
+		}
+
+		if len(kvs) != 1 || !bytes.Equal(kvs[0].Key, keys.SystemDBSpan.Key) {
+			return util.Errorf("expected only %s entry in SystemDB: %+v", keys.SystemDBSpan.Key, kvs)
```
nit: I think you want %q for the key, and the error string could be more precise in mentioning that the left-hand side is just a key while the right hand side is a slice of key-values

tschottdorf replied:
done.

tamird commented on storage/replica.go:
```diff
@@ -1571,16 +1571,28 @@ func (r *Replica) resolveIntents(ctx context.Context, intents []roachpb.Intent)
 	wg.Wait()
 }
 
+var errSystemDBIntent = fmt.Errorf("must retry later due to intent on SystemDB")
```
nit: errors.New but LGTM either way

tschottdorf replied:
done.

tamird commented on storage/replica_test.go:
```diff
@@ -3302,3 +3302,47 @@ func TestReplicaQuiesce(t *testing.T) {
+	v := roachpb.MakeValueFromString("foo")
+	// Create a transaction. We don't write a txn record, so pushing this will
+	// succeed and abort the intent we write here.
+	txn := newTransaction("test", []byte("a"), 1, roachpb.SNAPSHOT, rng.store.Clock())
```
oops, this is still SNAPSHOT

tschottdorf replied:
what a liar I am. Well, now done for reals.

tbg committed:
deal with intents on SystemDB

The previous code did not handle these. Since a leftover intent on
this table may only ever be touched by the code changed in this PR,
the previous version would stop gossipping the schema altogether.

With this change, when intents are encountered, an error is returned
but the intents are handled. This causes them to be cleaned up when
they are abandoned, but does not interfere with active writers. The
hope is that on the next iteration, the read can succeed. It is
going to be in our interest to minimize contention on that table.
